File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [0]*n
        pref = [0]*n
        suff = [0]*n

        pref[0] = suff[n-1] = 1

        # nums = [a,b,c,d]
        # 1, 1*a, 1*a*b, 1*a*b*c
        # 1*b*c*d, 1*c*d, 1*d, 1

        # b*c*d, a*c*d, a*b*d, a*b*c

        for i in range(1, n):
            pref[i] = nums[i-1] * pref[i-1]
        for i in range(n-2, -1, -1):
            suff[i] = nums[i+1] * suff[i+1]
        for i in range(n):
            res[i] = pref[i] * suff[i]
        
        return res

File: Data Structures & Algorithms/jump-game-ii/submission-0.java
class Solution {
    public int jump(int[] nums) {
        int res, l, r;
        res = l = r = 0;

        while (r < nums.length - 1) {
            int farthest = 0;
            for (int i = l;i < r + 1;i++) {
                farthest = Integer.max(farthest, i + nums[i]);
            }
            l = r + 1;
            r = farthest;
            res += 1;
        }
        return res;
    }
}
